greenmail-core/src/main/java/com/icegreen/greenmail/Managers.java

```diff
@@ -4,21 +4,104 @@
  */
 package com.icegreen.greenmail;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+
+import com.icegreen.greenmail.configuration.GreenMailConfiguration;
+import com.icegreen.greenmail.filestore.UncheckedFileStoreException;
 import com.icegreen.greenmail.imap.ImapHostManager;
 import com.icegreen.greenmail.imap.ImapHostManagerImpl;
 import com.icegreen.greenmail.smtp.SmtpManager;
-import com.icegreen.greenmail.store.InMemoryStore;
+import com.icegreen.greenmail.store.Store;
 import com.icegreen.greenmail.user.UserManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * @author Wael Chatila
  * @version $Id: $
  * @since Jan 27, 2006
  */
 public class Managers {
-    private ImapHostManager imapHostManager = new ImapHostManagerImpl(new InMemoryStore());
-    private UserManager userManager = new UserManager(imapHostManager);
-    private SmtpManager smtpManager = new SmtpManager(imapHostManager, userManager);
+    final Logger log = LoggerFactory.getLogger(Managers.class);
+
+    private ImapHostManager imapHostManager = null;
+    private UserManager userManager = null;
+    private SmtpManager smtpManager = null;
+    private Store storeToUse = null;
+
+    /**
+     * Public constructor wihtout Startup Configuration
+     *
+     * @deprecated Please use the constructor with a valid Startup Configuration
+     */
+    public Managers() {
+        this(new GreenMailConfiguration());
+    }
+
+    /**
+     * Public constructor with startupConfiguration
+     *
+     * @param startupConfig - the startup configuration
+     */
+    public Managers(GreenMailConfiguration startupConfig) {
+        log.info("Starting managers with the following startup configuration:");
+        startupConfig.logConfiguration();
+
+        // Depending on the startup confifguration, construct the correct Store class using Reflection:

```

Comment:
I had this as a switch-statement before, but then I was thinking about how I can I make the code in a way that subsequent changes do not touch so many files (I did not like my change to be so huge, but I really tried to leave it as small as possible)... 

Why? For example, now we have 2 stores. I think it would be possible that someone wants to have a third store (e.g. a filestore with mbox-format)... Either the guy who implements this additional store has to change two locations in the code (the location where the store is constructed and the store itself, and the guy who wants to add a store maybe does not know that the construction happens only at one location, so he has to know the whole source code for just adding a new feature). It would be more stable to have this one only at one place.

There is a default store implementation, and this default implementation can be overwritten by properties. I don't see any other way other than using Reflection is such a scenario where this should be generic and stable. This is then kind of a plugin system where classes for the store-implementation can be configured from outside.

But of course, if this is not really needed. I can program an if/then statement here, or a switch-statement. It makes this part of the source code simpler, but less generic and IMHO less open.

==> I will change it: No use of reflection, but use of hard-coded name of implementation classes.